src/python/pants/build_graph/mirrored_target_option_mixin.py

```diff
@@ -0,0 +1,92 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from abc import abstractmethod, abstractproperty
+
+from future.utils import text_type
+
+from pants.util.memo import memoized_property
+from pants.util.meta import AbstractClass
+from pants.util.objects import datatype
+
+
+class MirroredTargetOptionDeclaration(AbstractClass):
+  """An interface for operations to perform on an option which may also be set on a target."""
+
+  @abstractproperty
+  def is_flagged(self):
+    """Whether the option was specified on the command line."""
+
+  @abstractmethod
+  def extract_target_value(self, target):
+    """Get the value of this option from target. May return None if not set on the target."""
+
+  @abstractproperty
+  def option_value(self):
+    """Get the value of this option, separate from any target."""
+
+  def get_mirrored_scalar_option_value(self, target):
+    # Options specified on the command line take precedence over anything else.
+    if self.is_flagged:
+      return self.option_value
+
+    # Retrieve the value from the target, if set.
+    target_setting = self.extract_target_value(target)
+    if target_setting is not None:
+      return target_setting
+
+    # Otherwise, retrieve the value from the environment/pants.ini/hardcoded default.
+    return self.option_value
+
+
+class OptionableMirroredOptionDeclaration(datatype([
+    'options',
+    ('option_name', text_type),
+    'accessor',
+]), MirroredTargetOptionDeclaration):
+
+  @property
+  def is_flagged(self):
+    return self.options.is_flagged(self.option_name)
+
+  def extract_target_value(self, target):
+    return self.accessor(target)
+
+  @property
+  def option_value(self):
+    return self.options.get(self.option_name)
+
+
+class MirroredTargetOptionMixin(AbstractClass):
+  """Get option values which may be set in this subsystem or in a Target's keyword argument."""
```

Comment:
Expanding this pydoc with a usage example would be good.

Reply:
Done!